platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/RepositoryChainComponentMetaDataResolver.java

```diff
@@ -59,7 +72,19 @@ public void resolve(ComponentIdentifier identifier, ComponentOverrideMetadata co
             throw new UnsupportedOperationException("Can resolve meta-data for module components only.");
         }
 
-        resolveModule((ModuleComponentIdentifier) identifier, componentOverrideMetadata, result);
+        try {
+            CalculatedValueContainer<BuildableComponentResolveResult, ?> metadataValueContainer =
+                metadataValueContainerCache.get((ModuleComponentIdentifier) identifier, () -> createValueContainerFor(identifier, componentOverrideMetadata));
```

Comment:
🤔 I didn't notice this before. What's this `componentOverrideMetadata` doing?

With the ProducerGuard, it looks like we were (pessimistically) locking for a component identifier, even if `componentOverrideMetadata` would cause us to do something different.  Maybe only `isChanging` is meaningful?

So you could have two threads that come in with the same component identifier and different componentOverrideMetadata...
- In the old code, we would block, but look at both componentOverrideMetadata
- In the new code, we would block, but only use the first componentOverrideMetadata. 

Can you look into this?

Reply:
Per our conversation, I'm not aware of any scenarios where this could actually occur.  I've added some tests that pin down the behavior for client module dependencies and have verified that the behavior is consistent before and after this change.  If we discover any scenarios where this _can_ happen, we'll explore that and formulate a strategy to fix.